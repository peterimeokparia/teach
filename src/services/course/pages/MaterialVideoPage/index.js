import { 
createRef, 
PureComponent } from 'react';

import { 
connect } from 'react-redux';

import { 
uploadVideos } from 'services/course/helpers/ServerHelper';

import {
recordingStatusRecordingStarted,
recordingStatusRecordingStopped } from 'services/course/actions/video';

import { 
saveAs } from 'file-saver';

import VideoCallIcon from '@material-ui/icons/VideoCall';
import DeleteIcon from '@material-ui/icons/Delete';
import CancelIcon from '@material-ui/icons/Cancel';
import ScreenShareIcon from '@material-ui/icons/ScreenShare';
import './style.css';

class MaterialVideoPage extends PureComponent {

  constructor(props){
    super(props);
    this.state = {
      capture: false,
      screenSharing:false,
      cameraOn:false,
      mediaStream: null,
      selectedElement: null,
      selectedElementToDelete: null
    };
    this.videoRef = createRef();
    this.enableScreenSharing = this.enableScreenSharing.bind(this);
    this.enableCamera = this.enableCamera.bind(this);
    this.setCapture = this.setCapture.bind(this);
    this.setTypeOfStreamToCapture = this.setTypeOfStreamToCapture.bind(this);
    this.startCapture = this.startCapture.bind(this);
    this.stopCapture = this.stopCapture.bind(this);
    this.handleSelectedElement =this.handleSelectedElement.bind(this);
  };

  theStream;
  theRecorder;
  recorder;
  url;
  blob;
  recordedChunks = [];

  componentDidMount = () => {
  };

  componentDidUpdate = () => {
    this.cancelIcon = this.props?.videoMeta?.exitVideoCallIcon && 
       this.props?.videoMeta?.exitVideoCallIcon( this.state.capture, this.props.element?._id, this.state.selectedElement?._id );
  };

  componentWillUnmount = () => { 
  };

  requestedMediaOptions = {
    video: {
      cursor: "always"
    },
    audio: { 
      echoCancellation: true,
      noiseSuppression: true
    }
  };

  requestedVideoOptions = {
    video: {
      noiseSuppression: true,
    },
    audio: { 
      echoCancellation: true,
      noiseSuppression: true
    }
  };

  requestedAudioOptions = {
    audio: { 
      echoCancellation: true,
      noiseSuppression: true
    }
  };

  setCapture = () => {
   this.setState( { capture: !this.state.capture } );
  }; 

  closeVideo = () => {
    if ( this.state.capture ) {
      this.resetAfterCapture();
    }
    this.setState( { capture: false } );
    this.props.recordingStatusRecordingStopped();
  };

  toggleRecordOnOff = () => {  
    this.setCapture();
    if ( this.state.capture ) {
       this.resetAfterCapture();
    } else {
        this.startCapture();       
    }
  };

  resetAfterCapture = () => {
    this.stopCapture();

    if ( this.state.screenSharing ) {
      this.setState( { screenSharing: false } );
    }

    if ( this.state.cameraOn ) {
      this.setState( { cameraOn: false } );
    }
  };

  setTypeOfStreamToCapture = () => {
     this.setState( { screenSharing: !this.state.screenSharing } );
  };

  toggleTypeOfStremCapture = () => {
     this.setTypeOfStreamCapture();
     if ( this.state.screenSharing ) {
        this.enableScreenShare();     
     } else {
        this.enableCameraStream();
     }
  };

  startCapture = () => {
    this.props.extendedMeetingSettings && this.props?.setVideoModalMode(true);
    this.props.extendedMeetingSettings && this.props?.resetAllStartSettings();
    this.props.extendedMeetingSettings && this.props?.recordingStatusRecordingStarted();
    return "Record";
  };

  stopCapture = () => {
    if (this.state.capture && this.theRecorder && this.theStream){ 
        this.theRecorder.stop();
        this.theStream.getTracks()?.forEach( track => track.stop() );
        this.blob = new Blob(this.recordedChunks, {type: "video/webm"});
        this.url =  (URL || window.webkitURL).createObjectURL(this.blob);
        
    let videoData = {
        videoBlob: this.blob,  
        objectId: this.props?.videoMeta?.objectId,  
        videoMetaData: this.props?.videoMeta?.videoMetaData, 
        videoName: this.props?.videoMeta?.videoName };

      uploadVideos( videoData, this.props?.videoMeta?.videoMetaDataExternalId, this.props?.videoMeta?.videoNamePrefix ); 
      saveAs(this.url, "recording.webm");
      this.theStream = null;
      this.props.extendedMeetingSettings && this.props.resetAllStopSettings();
    }
      this.props.extendedMeetingSettings &&  this.props?.setVideoModalMode(false);
      this.props.extendedMeetingSettings && this.props?.recordingStatusRecordingStopped();      
      this.setState({ capture: false });
      this.setState({ cameraOn: false });
  };

  saveRecording = ( selectedElement ) => {
    this.props?.saveRecording( selectedElement );
  };

  enableScreenSharing = () => {
  if ( ( this.state.capture ) && ( this.state.screenSharing ) ) {
    this.stopCapture();
    this.setState({ screenSharing: false });
    this.setState({ capture: false });
    this.saveRecording( this.state.selectedElement );
    return;
  }
   navigator.mediaDevices.getDisplayMedia(this.requestedMediaOptions)
    .then(screen  => {
      this.addMicToTrack( screen, this.state.screenSharing );
    }).catch(e => { 
      console.error('getUserMedia() failed: ' + e); 
      this.props.extendedMeetingSettings && this.props?.resetAllStopSettings();
    }); 
    if ( this.state.cameraOn ) {
      this.setState( { cameraOn: false } );
    }
    this.setState( { screenSharing: true } );
  };

  enableCamera = () => {   
    if ( ( this.state.capture ) && ( this.state.cameraOn ) ) {
      this.stopCapture();
      this.setState({ cameraOn: false });
      this.setState({ capture: false });
      this.saveRecording( this.state.selectedElement );
      return;
    } 
    navigator.mediaDevices.getUserMedia(this.requestedVideoOptions).then(screen  => {
        this.addMicToTrack( screen, false );
    }).catch(e => { 
        console.error('getUserMedia() failed: ' + e); 
        this.props.extendedMeetingSettings && this.props.resetAllStopSettings();
    }); 
    if ( this.state.screenSharing ) {
      this.setState( { screenSharing: false } );
    }
    if ( this.state.capture ) {
      this.setState( { cameraOn: true } );
    }   
  };

  addMicToTrack = ( screen, screenSharingEnabled ) => {
    navigator.mediaDevices.getUserMedia(this.requestedAudioOptions).then(function(mic) {
      let tracks = mic.getTracks();
      let track = tracks[0];

      if ( screen && track ) {
        if ( screenSharingEnabled ) {
           screen.addTrack( track );
        }
        getStream(screen);
      }
    });

    let getStream = stream => {
      this.theStream = stream;
      let video = this.videoRef?.current;

      if ( video ) {
        video.srcObject = stream;
        video.play();
      }
      
      let options = {mimeType: 'video/webm;codecs=vp9,opus'};

      if (! MediaRecorder.isTypeSupported(options.mimeType)) {
          console.error(`${options.mimeType} is not supported`);
          options = {mimeType: 'video/webm;codecs=vp8,opus'};
        if (! MediaRecorder.isTypeSupported(options.mimeType)) {
          console.error(`${options.mimeType} is not supported`);
          options = {mimeType: 'video/webm'};
          if (! MediaRecorder.isTypeSupported(options.mimeType)) {
            console.error(`${options.mimeType} is not supported`);
            options = {mimeType: ''};
          }
        }
      }
  
      try {
        console.log('MediaRecorder', stream);
        this.recorder = new MediaRecorder(stream, options);
      } catch (e) {
        console.error('Exception while creating MediaRecorder: ' + e);
        return;
      }
      
      this.theRecorder = this.recorder;
  
      if ( this.recordedChunks ){
          this.recordedChunks = [];
      }
      this.recorder.ondataavailable = (event) => { this.recordedChunks.push(event?.data); };
      this.recorder.start(100);
    };
  };
  
  handleSelectedElement = ( selected ) => {
    this.setState({ selectedElement: selected });
    this.toggleRecordOnOff();
  };

  handleSelectedElementToDelete = ( selected ) => {
    this.saveRecording({ ...selected, videoUrl:''}); 
  };

  handleCanPlay = () => {       
    this.videoRef.current.play();
  };

  render(){
        return (
            <>
            <span className="">  
               <span>  
              { 
                <span className={"videoSectionMain"}> 
                  <VideoCallIcon 
                      style={  this.props?.videoMeta?.videoCallIconMain( this.state.capture , this.props.element?._id, this.state.selectedElement?._id  )}
                      className={ ( this.state.capture && (  this.props.element?._id === this.props.selectedElement?._id )) ? "comment-round-button-3" : "comment-round-button-4" }
                      onClick={() => this.handleSelectedElement(  this.props.element )} 
                  /> 
                  <DeleteIcon 
                      style={ this.props?.videoMeta?.deleteIconStyle(  this.state.capture, this.props.element?._id, this.state.selectedElement?._id )  }
                      className="comment-round-button-3"
                      onClick={() => this.handleSelectedElementToDelete( this.props.element ) }
                  />
                </span>
              }      
              {
                  <div 
                  className={ ( this.state.capture ) ? this.props?.videoMeta?.videoSectionClassNameRecording : this.props?.videoMeta?.videoSectionClassNameRecordingStopped  }
                >
                  {( this.state.capture && this.props?.videoMeta?.displayVideoPlayer ) && 
                        <video
                          controls 
                          autoPlay="true"
                          className="video"
                          ref={ this.videoRef }
                          onCanPlay={ this.handleCanPlay } 
                          onBlur={ () => this.closeVideo() }
                      />
                  } 
                </div>        
              }
              { this.state.capture &&  (   
                <div className={this.props?.videoMeta?.videoSectionCallOut}>   
                  {      
                    <span> 
                      <CancelIcon 
                          style={ this.props?.videoMeta?.exitVideoCallIcon( this.state.capture, this.props.element?._id, this.state.selectedElement?._id ) }
                          className={ ( this.state.capture ) ? ( this.state.cameraOn ) ? "comment-round-button-3" : "comment-round-button-2" : "comment-round-button-4" }
                          onClick={ this.closeVideo }
                      />
                      <VideoCallIcon 
                          style={ this.props?.videoMeta?.videoCallIcon( this.state.capture, this.props.element?._id, this.state.selectedElement?._id  ) }
                          className={ ( this.state.capture ) ? ( this.state.cameraOn ) ? "comment-round-button-3" : "comment-round-button-2" : "comment-round-button-4" }
                          onClick={ this.enableCamera }
                      />
                       <ScreenShareIcon 
                          style={ this.props?.videoMeta?.shareScreenIcon( this.state.capture, this.props.element?._id, this.state.selectedElement?._id ) }
                          className={ ( this.state.capture ) ? ( this.state.screenSharing ) ? "comment-round-button-3" :  "comment-round-button-2" : "comment-round-button-4" }
                          onClick={ this.enableScreenSharing }
                        />
                    </span>
                  } 
                </div>
              )}
              </span> 
        </span>
        </>
        );
      }; 

}

const mapDispatch = {
  recordingStatusRecordingStarted,
  recordingStatusRecordingStopped,
};

const mapState = ( state, ownProps ) => {
  return {
    courses: Object.values( state?.courses?.courses ),
    hasRecordingStarted: state.hasRecordingStarted.hasRecordingStarted
  };
};

export default connect( mapState, mapDispatch )( MaterialVideoPage );





