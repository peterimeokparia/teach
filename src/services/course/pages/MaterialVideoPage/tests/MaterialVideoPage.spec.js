import React from 'react';
import MaterialVideoPage  from 'services/course/pages/MaterialVideoPage';
import renderer from 'react-test-renderer';
import thunk from "redux-thunk"; 
import * as reactRedux from 'react-redux';
import { Provider, useSelector } from "react-redux";
import { applyMiddleware } from 'redux'; 
import configureStore from 'redux-mock-store';

export const videoCallIconMain = ( capturingVideo, id, selectedId ) => {
    return {
      fontSize: 60,
      "visibility": ( capturingVideo && id === selectedId ) ? "hidden" : "visible",
    };
  };

  export const videoCallIcon = ( capturingVideo, id, selectedId ) => {
    return {
      fontSize: 60,
      "visibility": ( capturingVideo && id === selectedId ) ? "hidden" : "visible",
    };
  };

  export const shareScreenIcon = ( capturingVideo, id, selectedId  ) => {
    return {
      fontSize: 55,
      position: "relative",
      visibility: ( capturingVideo && id === selectedId ) ? "visible" : "hidden"
    };
  };
  
  export const exitVideoCallIcon = (capturingVideo, id, selectedId  ) => {
    return {
      fontSize: 60,
      position: "relative",
      "visibility": ( capturingVideo && id === selectedId ) ? "visible" : "hidden",
    };
  };

  export const onlineQuestionVideoDeleteIconStyle = (capturingVideo, id, selectedId  ) => {
    return {
      fontSize: 60,
      position: "relative",
      "visibility": ( capturingVideo && id === selectedId ) ? "visible" : "hidden",
    };
  };

  let element = { _id: "10000004"}

 let videoMeta = element  => { 
    return {
      videoCallIconMain:videoCallIconMain,
      deleteIconStyle: onlineQuestionVideoDeleteIconStyle,
      videoCallIcon,
      shareScreenIcon,
      exitVideoCallIcon,
      videoNamePrefix: 'OnlineQuestionVideoMarkDownEditors', 
      recordButtonText: 'Record Question',
      displayMaterialButton: true,
      videoSectionClassNameRecording: "mainVideoSection-recording",
      videoSectionClassNameRecordingStopped: "mainVideoSection-recordingStopped",
      videoSectionClassNameNoRecording: "mainVideoSection-recordingStopped", 
      videoClassName: ( element?.videoUrl === ""  ) ? "videoPlayer" : "",
      exitVideoCallIconPageName: "OnlineListItems",
      videoSectionCallOut: "videoSectionCallOut",
      videoMetaData: { inputFieldId: element?._id, currentQuestion: element, name: element?._id?.toString() },
      videoName: `${element?._id}_${element?._id}_${element?._id}_${element?.type}`,
      videoMetaDataExternalId:'name',
      buttonClassName: `toggle-stage-btns${( true ) ? "-show" : "-hide"}`, 
      objectId: element?._id, 
      displaySavedRecording: true
    }};

jest.mock('../../../Api');

const mockStore = configureStore([]);
  
let operatorBusinessName= "boomingllc";
let selectedUserId= "PERSON010010100000047";

describe('MaterialVideoPage', () =>  {  

 let  store = mockStore({
      operators: { operators: {_id: "5fcb0e19fd5e0117dc09dcfa", operatorBusinessName} },
      notifications: { pushNotificationSubscribers : {_id:"5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" } },
      users: { users: {_id: "PERSON5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      courses: { courses: {_id: "COURSE5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      grades: { grades: {_id: "COURSE5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      classrooms: { classrooms: {displaySideBarDropDown: false }},
      sessions: { sessions: {_id: "SESSIONS5fcb0e19fd5e0117dc09dcfa", numberOfSessions: 9, totalNumberOfSessions: 1, status: true } },
      lessons: { lessons: { selectedLessonFromLessonPlanDropDown: "Test Lesson" } },
      courses: { courses: { selectedCourseFromLessonPlanCourseDropDown: "Test Course" }  },
      meetings: { meetings: {_id: "MEETING5fcb0e19fd5e0117dc09dcfa", numberOfSessions: 9, totalNumberOfSessions: 1, status: true }  },
      hasRecordingStarted: { hasRecordingStarted: true}
  }, applyMiddleware(thunk));

  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
  const element = {}
 
  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  it('renders as expected', async () => {

    const dispatch = jest.fn();
    useDispatchMock.mockReturnValue(dispatch);
    
    const saveRecording = () => {}
    const resetAllStartSettings = () => {}
    const resetAllStopSettings = () => {}
    function setVideoModalMode( modal ) {
        console.log('test')
    };


    const tree = renderer.create(
      <Provider store={store}>
        <MaterialVideoPage 
                 videoMeta={videoMeta()} 
                 element={element}
                 saveRecording={saveRecording} 
                 resetAllStartSettings={resetAllStartSettings}
                 resetAllStopSettings={resetAllStopSettings}
                 setVideoModalMode={(modal) => setVideoModalMode(modal)}
                 videoModalMode={true}
                 extendedMeetingSettings={false}
        //   selectedUserId={selectedUserId}
        //   operatorBusinessName={operatorBusinessName } 
      /> 
    </Provider>
    );
      //console.log( happy - tree  - no sad trees );
      expect(tree.toJSON()).toMatchSnapshot();
  })
});




