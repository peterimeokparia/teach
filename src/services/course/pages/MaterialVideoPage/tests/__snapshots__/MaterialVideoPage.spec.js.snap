// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialVideoPage renders as expected 1`] = `
<span
  className=""
>
  <span>
    <span
      className="videoSectionMain"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root comment-round-button-4"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "fontSize": 60,
            "visibility": "visible",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M17 10.5V7c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1v10c0 .55.45 1 1 1h12c.55 0 1-.45 1-1v-3.5l4 4v-11l-4 4zM14 13h-3v3H9v-3H6v-2h3V8h2v3h3v2z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root comment-round-button-3"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "fontSize": 60,
            "position": "relative",
            "visibility": "hidden",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    <div
      className="mainVideoSection-recordingStopped"
    />
  </span>
</span>
`;
